### ISSUE TITLE
Editable class not working inside ng-repeat directive

### ISSUE THREAD

**Issue by alchemistgo87** [yabwe/medium-editor#739]
```javascript
<div ng-repeat="subSection in aboutSection.subSections">
<div class="editable">this is editable content</div>
</div>				
```
In this I am not able to edit "this is editable content" while its working outside of ng-repeat. I am using latest 5.5.0 version.

**Comment by nmielnik** [comment]
Would you be able to give a bit more clarify as to what you're doing and how you're instantiating MediumEditor? Something like a JSFiddle would be extremely helpful so we can get a better idea of what is going on and what problem you are seeing.

**Comment by alchemistgo87** [comment]
Her is the JSFiddle:
https://jsfiddle.net/L5j2ym9e/2/

**Comment by nmielnik** [comment]
@alchemistgo87 I believe the issue here is that you are including the script tag which initializes MediumEditor inside the `<head>`.  At this point, the `.editable` element doesn't exist on the page.

I've updated your JSFiddle and was able to get things working by including the `<script>` tag after your HTML: https://jsfiddle.net/L5j2ym9e/7/

You can try either including the `<script>` tag further within the body, or you can wait until the document has loaded, and then trigger your script to initialize `MediumEditor`.

If this doesn't resolve the issue, please re-open and let us know what you're seeing.  Thanks!

**Comment by alchemistgo87** [comment]
This is the actual jfiddle: https://jsfiddle.net/L5j2ym9e/9/
I will explain what I am doing. There is list of sections and subsections which user can create dynamically by pressing add button. On pressing add button input boxes are created inside the ng-repeat directive as I add an empty item in the list of subsections on pressing of addbutton. Now here is a textarea inside the subsection. This textarea is not showing any editor. But I am able to get an editor if I use a textarea outside of ng-repeat directive.

**Comment by nmielnik** [comment]
So I'm able to muck around with the jsfiddle as there is some exception getting thrown by Angular.

I would add a line into your code wherever you're instantiating MediumEditor to query how many `.editable` elements are available on the page at that time.

Something like:
```js
console.log(document.querySelectorAll('.editable').length);
editor = new MediumEditor('.editable', { } );
```

What is likely happening is that the elements do not exist at the time MediumEditor is instantiated.  Especially if the user can dynamically add more of these, you would need to make sure the elements exist before creating a new instance of MediumEditor.

**Comment by alchemistgo87** [comment]
You are right, elements do not exist during the creation of a new instance of MediumEditor. So, it means whenever I am dynamically adding a new element, I need to re-initialize an instance of MediumEditor?

**Comment by alchemistgo87** [comment]
```javascript
$scope.addSubSection = function (aboutSection) {
      	$scope.editor = new MediumEditor('.editable', {
			buttonLabels: 'fontawesome'
		});
      	
      	aboutSection.subSections.push({heading:'', sequence: '', description:'', mediaList:[], mediaPosition:''});
      	
        }
```
Now on every call of addSubSection I have re-instantiated MediumEditor. But editor comes only after I call addSubSection 2 times. On the second call the editor becomes visible in the first subsection and so on. What am I doing wrong here?

**Comment by nmielnik** [comment]
My guess for this would be that you need to swap the two lines within your `addSubSection` method.

Make sure you do all of your html manipulation first (ie creating the subsection elements), and then call `new MediumEditor()`.  I'm not familiar with angular, but I'm assuming that the call to `aboutSection.subSections.push()` will actually put the HTML into the page?  If so, then this would work:

```javascript
$scope.addSubSection = function (aboutSection) {
    aboutSection.subSections.push({heading:'', sequence: '', description:'', mediaList:[], mediaPosition:''});

    $scope.editor = new MediumEditor('.editable', {
        buttonLabels: 'fontawesome'
    });
}
```

Also, if this line of code is getting called multiple times, that means each time you instantiate `MediumEditor` it will look for all `.editable` elements on the page.  This can be problematic as multiple instances of MediumEditor will attempt to manage the same element. (ie, the first created `.editable` element will then be managed by ever new instance you create, since you're passing `.editable` as the selector).

It's perfectly fine to create a new instance of MediumEditor each time you create a new element, my org does this as well since we allow users to dynamically add elements.

In this case, I would recommend either using a unique id as the selector you pass to MediumEditor, or get a reference to the specific element itself and pass it to `MediumEditor`:

```javascript
$scope.addSubSection = function (aboutSection) {
    aboutSection.subSections.push({heading:'', sequence: '', description:'', mediaList:[], mediaPosition:''});

    // option 1: get a reference to the element you just created
    var textareaRef =  // If you can get a reference to the textarea you just created, get it here
    $scope.editor = new MediumEditor(textareaRef, {
        buttonLabels: 'fontawesome'
    });

    // option 2: if there's a unique identifier on the element you just created, use that
    $scope.editor = new MediumEditor('#textarea1', {
        buttonLabels: 'fontawesome'
    });
}
```

Hope that helps

**Comment by alchemistgo87** [comment]
I got the answer why it was not working, needed to add "$scope.$apply();" after  pushing the element into the array. But now another problem has occured:

Using <textarea> with class="editable medium-editor-textarea" disrupts binding with ng-model while using <textarea> without the editable class is working. So how to ensure binding with ng-model while keeping the class editable?

**Comment by nmielnik** [comment]
The first thing I can tell you is that the `medium-editor-textarea` class is not needed for MediumEditor to work.  You don't even need the `editable` class on your `<textarea>` as you can pass in any kind of selector to MediumEditor (ie id, data-attribute, etc.).  You can also just pass the element itself in when you create the instance of MediumEditor.  So, neither of those classes are needed to work.

I'm not too familiar with Angular, so I would recommend you dig into what kind of changes to an element would disrupt the binding.

What I can tell you is happening internally within MediumEditor, is that when it encounters a `<textarea>` element, it will add a class name to the `<textarea>` (medium-editor-hidden) which should cause the `<textarea>` to be hidden.  It will also add a `medium-editor-textarea-id` attribute to the `<textarea>` element, which allows the code to maintain a reference between that `<textarea>` and the editor element it's going to create.

It then creates a `<div>` which is `contenteditable` and links the two elements together by monitoring any change to the `<div>` and updating the contents of the `<textarea>`.  It also copies a bunch of attributes that MediumEditor cares about from the `<textarea>` to the `<div>`.

When the editor is destroyed, the `<div>` is removed and the `<textarea>` is displayed again by removing the `medium-editor-hidden` class.

**Comment by alchemistgo87** [comment]
Tried all the methods, even added an on ng-repeat render finish event.  But still binding is disrupted.  Not able to find out the solution.

**Comment by injectless** [comment]
does anyone have an answer for this?
at least, why is this happening?

```
 <md-tab ng-repeat="tab in tabs">
<div class="editable" data-placeholder="Type some text">
           Text here
 </div>
</md-tab>
```

It is also not working.

**Comment by injectless** [comment]
I found the solution.
You have to help the $scope listen for changes, and this is where $watch takes place.

```
$scope.$watch('selectedIndex', function (...) {
            var editor = new MediumEditor('.editable');

            $('.editable').mediumInsert({
                editor: editor,
                addons: {
                    images: {
                        uploadScript: null,
                        deleteScript: null,
                        captionPlaceholder: 'Type caption for image',
                        styles: {
                            slideshow: {
                                label: '<span class="fa fa-play"></span>',
                                added: function ($el) {
                                    $el
                                        .data('cycle-center-vert', true)
                                        .cycle({
                                            slides: 'figure'
                                        });
                                },
                                removed: function ($el) {
                                    $el.cycle('destroy');
                                }
                            }
                        },
                        actions: null
                    }
                }
            });
        });
```

**Comment by nmielnik** [comment]
Thanks @injectless!
